phoenix-core/src/main/java/org/apache/phoenix/expression/function/CollationKeyFunction.java

```diff
@@ -0,0 +1,230 @@
+
+			if (LOG.isDebugEnabled()) {
+				LOG.debug("Collation key bytes: " + VarBinaryFormatter.INSTANCE.format(collationKeyByteArray));
+			}
+
+			ptr.set(collationKeyByteArray);
+			return true;
+		} catch (ExpressionEvaluationException e) {
+			LOG.debug("ExpressionEvaluationException caught: " + e.getMessage());
+			return false;
+		}
+	}
+
+	private void initialize() {
+		String localeISOCode = getLiteralValue(1, String.class);
+		Boolean useSpecialUpperCaseCollator = getLiteralValue(2, Boolean.class);
+		Integer collatorStrength = getLiteralValue(3, Integer.class);
+		Integer collatorDecomposition = getLiteralValue(4, Integer.class);
+
+		if (LOG.isDebugEnabled()) {
+			StringBuilder logInputsMessage = new StringBuilder();
+			logInputsMessage.append("Input (literal) arguments:").append("localeISOCode: " + localeISOCode)
+					.append(", useSpecialUpperCaseCollator: " + useSpecialUpperCaseCollator)
+					.append(", collatorStrength: " + collatorStrength)
+					.append(", collatorDecomposition: " + collatorDecomposition);
+			LOG.debug(logInputsMessage);
+		}
+
+		Locale locale = LocaleUtils.get().getLocaleByIsoCode(localeISOCode);
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(String.format("Locale: " + locale.toLanguageTag()));
+		}
+
+		LinguisticSort linguisticSort = LinguisticSort.get(locale);
+
+		collator = BooleanUtils.isTrue(useSpecialUpperCaseCollator) ? linguisticSort.getUpperCaseCollator(false)
+				: linguisticSort.getCollator();
+
+		if (collatorStrength != null) {
+			collator.setStrength(collatorStrength);
+		}
+
+		if (collatorDecomposition != null) {
+			collator.setDecomposition(collatorDecomposition);
+		}
+
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(String.format("Collator: [strength: %d, decomposition: %d], Special-Upper-Case: %s",
+					collator.getStrength(), collator.getDecomposition(),
+					BooleanUtils.isTrue(useSpecialUpperCaseCollator)));
+		}
+	}
+
+	@Override
+	public PDataType getDataType() {
+		return PVarbinary.INSTANCE;
+	}
+
+	@Override
+	public String getName() {
+		return "COLLATION_KEY";
+	}
+
+	/**
+	 * Get the string for which the collation key needs to be calculated (the
+	 * first argument)
+	 */
+	private String getInputString(Tuple tuple, ImmutableBytesWritable ptr) {
+		Expression expression = getChildren().get(0);
+		if (!expression.evaluate(tuple, ptr)) {
+			throw new ExpressionEvaluationException(expression);
+		}
+		String inputString = (String) PVarchar.INSTANCE.toObject(ptr, expression.getSortOrder());
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("inputString: " + inputString);
+		}
+		return inputString;
+	}
+
+	private <T> T getLiteralValue(int childIndex, Class<T> type) {
+		Expression expression = getChildren().get(childIndex);
+		if (LOG.isDebugEnabled()) {
+			LOG.debug("child: " + childIndex + ", expression: " + expression);
+		}
+		if (!(expression instanceof LiteralExpression)) {
+			throw new ExpressionEvaluationException(expression);
+		}
```

Comment:
This if statement isn't necessary. Phoenix already checks this and throws if it's not the case based on your declaration of the argument as being a constant.